### ISSUE TITLE
Wish list

### ISSUE THREAD

**Issue by thetutlage** [Julien-R44/adonis-vscode-extension#22]
As we progress with v6, I have some ideas about how to make the extension even more awesome. Also, we might have to change a few things considering the changes in v6.

So this issue captures a **wish list**, **discusses some essential changes**, and can be used to further take requests from AdonisJS users.

## 🔧 Routes listing changes
Since routes controller binding in v6 can be done in many different ways, the `list:routes` command does its best to find the path to the controller file but may not be able to do it every time. Let's look at some examples.

#### Magic strings
No changes from v5. We are now using Node.js subpath imports to alias the controllers directory.

```ts
import router from '@adonisjs/core/services/router'
router.get('users', '#controllers/users_controller.index')
```

Output

```json
{
  "handler": {
	"type": "controller",
    "moduleNameOrPath": "#controllers/users_controller",
    "method": "index",
  },
}
```

#### Binding controller class
In the following example, there is no way for the `list:routes` command to know the import path for the controller. All it knows is the class name.

```ts
import router from '@adonisjs/core/services/router'
import UsersController from '#controllers/users_controller'

router.get('users', [UsersController, 'index'])
```

Output

```json
{
  "handler": {
	"type": "controller",
    "moduleNameOrPath": "UsersController",
    "method": "index",
  },
}
```

#### Lazy loading controller
In the following example, we parse the `import` method from the function that lazy loads the controller. The process is straightforward, call the `toString()` method on the function and parse the `import` statement. 

```ts
import router from '@adonisjs/core/services/router'
const UsersController = () => import('#controllers/users_controller')

router.get('users', [UsersController, 'index'])
```

Output (same as magic string)

```json
{
  "handler": {
	"type": "controller",
    "moduleNameOrPath": "#controllers/users_controller",
    "method": "index",
  },
}
```

Right now, the VSCode extension displaying routes in the activity sidebar takes the user to the controller file based on the controller path. However, in v6, we will have instances where we will not have the controller file path.

I am still trying to figure out a good answer on how to handle it. So let's keep it open for discussion.

## 💌 Refactor standard import to lazy import

This will be a great addition if we can support it. First, look at the examples of `Standard import` and `Lazy import`.

```ts
// Standard Import
import UsersController from '#controllers/users_controller'
```

```ts
// Lazy import
const UsersController = () => import('#controllers/users_controller')
```

Lazy import is preferred, as it keeps the boot phase of the application fast. However, typing the complete import statement and variable assignment by hand is tiring.

VSCode has auto imports, so if I type `Users`, it will show me a suggestion to import the `UsersController` from the controller file. First, however, the auto import will write a **Standard import** statement. 

So, adding a command to our extension that converts the **Standard import** to **Lazy import** will be freaking cool.

## 💌 Go to component from Edge tags

Right now, in the edge template files, I can jump to template references inside the `includes` and the `component` tags. However, we also allow components to be used as tags. For example:

Instead of writing

```html
@component('button')
@end
```

I can write

```html
@button()
@end
```

So it will be nice if clicking on the `@button` can take me to the component file. If needed, we can expose some APIs from Edge or AdonisJS to make it happen. But I don't know how to start approaching it.

## 💌 Bunch of snippets

I have a rough collection of snippets in mind that I would love to see and use. Also, we can take feedback from the community and ask for snippets they would love to use.

I can take care of doing a PR for snippets.

## 🔧 Does Edge syntax highlighting need improvements?

I remember we used some existing template syntax files inside the AdonisJS extension, and that syntax file has some rough edges like it does not highlight the `@end` tag correctly. It does highlight `@end<something>`, but not the standalone `@end` tag.

For me, just fixing the one quirk is enough. Unless Julien you have some plans to write it from scratch.

**Comment by Julien-R44** [comment]
about "Refactor standard import to lazy import"

I think the ideal would be to have an eslint rule for this. For example, we could develop a **auto-fixable** eslint rule "prefer-routing-with-lazy-controller-import".

```js
// .eslintrc.js

export default {
  "rules": {
    "adonisjs/prefer-routing-with-lazy-controller-import": "warning"
  }
}
```

And with that, it would automatically convert the import to lazy-import when we save the file. 

Otherwise, I need to check something else. I need to see if it's possible via VScode API to add items to the import suggestions. In that case, we'd have a choice between "UsersController" and "UsersController ( Lazy-import )". If this is possible, it would be a cool feature to have in addition to the eslint plugin

But if we have the eslint plugin, then I don't think we need to expose an extra VScode command to refactor to lazy-import. What do you think?

**Comment by thetutlage** [comment]
Yeah, having it as an ESLint rule does make sense. If someone does not want it, they can disable the rule

**Comment by Julien-R44** [comment]
> Right now, the VSCode extension displaying routes in the activity sidebar takes the user to the controller file based on the controller path. However, in v6, we will have instances where we will not have the controller file path.

Unfortunately it's very annoying and I don't think we have any solutions for that. The only thing I can do on my side, when I don't have the path import, is to do a kind of text search. So if I get :

```json
{
  "handler": {
	"type": "controller",
    "moduleNameOrPath": "UsersController",
    "method": "index",
  },
}
```

I can't even rely on file names. Because `UsersController` may be stored in a file `users_controller.ts` or `user.controller.ts` or whatever

So I have to search all the files content in the workspace and their default export, and look for a export default UsersController. I think it's a pretty heavy operation. And on top of that, it's not reliable, as you could very well have a default UsersController exported somewhere else.

Let's imagine we have a controller with a much less specific name, then we're screwed :

```
- app/
 - features/
   - posts/
     - controllers/
        - list_controller.ts > export default ListController
   - comments/
       - controllers/
         - list_controller.ts > export default ListController
```

a bit far-fetched, but possible. And in this case, it's impossible to tell which one we have to select

So it's quite annoying. I'll think about it some more, but right now I think we're screwed and don't have any solution unfortanely 😞

**Comment by thetutlage** [comment]
Can you try running the `list:routes` command with the v6 starter kit and see if the output there is helpful? I created this helper to get as accurate info as possible for the controller. https://github.com/adonisjs/core/blob/next/src/helpers/parse_binding_reference.ts#L13

**Comment by thetutlage** [comment]
One more thing. Does, VSCode has any API to search for symbols in a project? Because, if we have the controller class name (ie is a symbol), then we can search that symbol and get its filename.

Or, in the first iteration, we can just ignore routes with no module path and later think of a solution.

**Comment by Julien-R44** [comment]
Yeah I tried list:routes with the V6, and well, the only problem is with classic imports. For the rest, it should work, I just need to parse the import subpaths of the project defined in the package.json, and then I can easily find the file from `moduleNameOrPath`

VScode does offer an API for using symbol search, but the problem remains the same: you could very well have several controllers with the same name, so it's not very reliable. Example : 
![image](https://github.com/Julien-R44/adonis-vscode-extension/assets/8337858/2cc9dfe7-d2b4-4d22-900a-b491bdf85e7f)

But I'm thinking, maybe it's better to do that than to disable the feature altogether. 
I'll keep you updated

**Comment by thetutlage** [comment]
The chances of having two controllers with same name is less, because we have controller suffix too.

So, yeah lets use that API

**Comment by McSneaky** [comment]
Haven't gone thro whole issue yet, just noticed that two controllers with same name

It's quite common to have them in different folders for different versions
`v1/DatasetsController`, `v2/DatasetsController`, `v3/DatasetsController` etc
Would that fall under controller suffix? 🤔

**Comment by Julien-R44** [comment]
Okay so we have the routes support for v6 + edge components as tags : 

--- 

## Component as tags 

Works exactly like the rest, we have two new providers, one for links, the other for autocomplete:

![image](https://github.com/Julien-R44/adonis-vscode-extension/assets/8337858/35c3a7b2-487e-4f45-bf31-5c9274fd4e0a)

The autocomplete appears just after a CTRL + space, just like the rest. Very cool :) 

## Routes change 

Works as before, with a few additions:

- For V6 standard controller imports, I've set up that symbol searching thing we were talking about:

![image](https://github.com/Julien-R44/adonis-vscode-extension/assets/8337858/37497696-c3df-4b9a-9488-dcc11c644f36)

As you can see here:
- I have two controllers with the same name. Two UsersController classes.
- v1/users/show and v2/users/show are binded routes with standard imports. So there's no information about the module paths.
- When I click on the route in the TreeView, I'm asked to choose a controller to open. What we do here is search the workspace for all class symbols with the name UsersController. If there's only one, it's automatically opened without a picker. If several with the same name are found, this picker is displayed.

- A picker that lets you change the application for which routes are displayed in the TreeView. Useful when working with a monoreposit or several workspaces
![image](https://github.com/Julien-R44/adonis-vscode-extension/assets/8337858/4f883b08-e915-4c9b-b33b-56456d3302b2)

- I've removed the open in browser on non-GET routes. I also searched whether it was possible to open the route directly in Postman / Insomnia via a deeplink, but I couldn't find anything :/ 

---

Published as 1.1.1 on the marketplace 🤙

**Comment by Julien-R44** [comment]
I think we can close this! The pre-release is in pretty good state actually. I am thinking of moving it to stable release in the days/weeks to come

Thanks again for ideas and feedback @thetutlage !
